lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/IMRUDriver.cs

```diff
@@ -98,181 +128,457 @@ internal sealed class IMRUDriver<TMapInput, TMapOutput, TResult, TPartitionType>
+            {
+                if (_taskManager.AreAllTasksCompleted())
+                {
+                    Logger.Log(Level.Verbose, "Evaluator with Id: {0} failed but IMRU task is completed. So ignoring.", failedEvaluator.Id);
+                    return;
+                }
+                Logger.Log(Level.Info, "Evaluator with Id: {0} failed with Exception: {1} in retry # {2}.", failedEvaluator.Id, failedEvaluator.EvaluatorException, _numberOfRetryForFaultTolerant);
+
+                var isMaster = _evaluatorManager.IsMasterEvaluatorId(failedEvaluator.Id);
+                _evaluatorManager.RecordFailedEvaluator(failedEvaluator.Id);
+                _contextManager.RemoveFailedContextInFailedEvaluator(failedEvaluator);
+
+                switch (_systemState.CurrentState)
+                {
+                    case SystemState.WaitingForEvaluator:
+                        if (Recoverable())
+                        {
+                            if (isMaster)
+                            {
+                                Logger.Log(Level.Info, "Requesting a master Evaluator.");
+                                _evaluatorManager.RequestUpdateEvaluator();
+                            }
+                            else
+                            {
+                                _serviceAndContextConfigurationProvider.RemoveEvaluatorIdFromPartitionIdProvider(failedEvaluator.Id);
+                                Logger.Log(Level.Info, "Requesting mapper Evaluators.");
+                                _evaluatorManager.RequestMapEvaluators(1);
+                            }
+                        }
+                        else
+                        {
+                            Logger.Log(Level.Warning, "The system is not recoverable, change the state to Fail.");
+                            _systemState.MoveNext(SystemStateEvent.NotRecoverable);
+                            FailAction();
+                        }
+                        break;
+
+                    case SystemState.SubmittingTasks:
+                    case SystemState.TasksRunning:
+                        // set system state to ShuttingDown
+                        _systemState.MoveNext(SystemStateEvent.FailedNode);
+                        _taskManager.RecordTaskFailWhenReceivingFailedEvaluator(failedEvaluator);
+                        _taskManager.CloseAllRunningTasks(TaskManager.CloseTaskByDriver);
+
+                        // Push evaluator id back to PartitionIdProvider if it is not master
+                        if (!isMaster)
+                        {
+                            _serviceAndContextConfigurationProvider.RemoveEvaluatorIdFromPartitionIdProvider(failedEvaluator.Id);
+                        }
+
+                        TryRecovery();
+                        break;
+
+                    case SystemState.ShuttingDown:
+                        _taskManager.RecordTaskFailWhenReceivingFailedEvaluator(failedEvaluator);
```

Comment:
Ok, I added Fail case. In that case, it fails and we just log and do noting, 
For TaskCompleted state, it has been covered at the begging of the method. 
No more others.